BootService: Crashes if media is 'ejecting'

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/ad76bad2-3161-48c2-afc4-b9ab08216043

Also occurs on 2.9.6

Useful logs:

getExternalStorageState | ejecting
-- | --


```
java.lang.RuntimeException: Unable to create application com.ichi2.anki.AnkiDroidApp: android.database.sqlite.SQLiteCantOpenDatabaseException: unknown error (code 14): Could not open database
at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5405)
at android.app.ActivityThread.-wrap2(ActivityThread.java)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1541)
at android.os.Handler.dispatchMessage(Handler.java:102)
at android.os.Looper.loop(Looper.java:154)
at android.app.ActivityThread.main(ActivityThread.java:6123)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:867)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:757)
Caused by: android.database.sqlite.SQLiteCantOpenDatabaseException: unknown error (code 14): Could not open database
at io.requery.android.database.sqlite.SQLiteConnection.nativeOpen(Native Method)
at io.requery.android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:5)
at io.requery.android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:2)
at io.requery.android.database.sqlite.SQLiteConnectionPool.openConnectionLocked(SQLiteConnectionPool.java:2)
at io.requery.android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:4)
at io.requery.android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:2)
at io.requery.android.database.sqlite.SQLiteDatabase.openInner(SQLiteDatabase.java:2)
at io.requery.android.database.sqlite.SQLiteDatabase.open(SQLiteDatabase.java:3)
at io.requery.android.database.sqlite.SQLiteDatabase.openDatabase(SQLiteDatabase.java:6)
at io.requery.android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:16)
at io.requery.android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:3)
at io.requery.android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:1)
at com.ichi2.libanki.DB.<init>(DB.java:8)
at com.ichi2.libanki.Storage.Collection(Storage.java:5)
at com.ichi2.anki.CollectionHelper.getCol(CollectionHelper.java:6)
at com.ichi2.anki.services.BootService.onReceive(BootService.java:3)
at com.ichi2.anki.AnkiDroidApp.onCreate(AnkiDroidApp.java:29)
at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1046)
at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5402)
... 8 more
```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/services/BootService.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/services/BootService.java
@@ -35,8 +35,8 @@ public class BootService extends BroadcastReceiver {
             return;
         }
         // There are cases where the app is installed, and we have access, but nothing exist yet
-        if (CollectionHelper.getInstance().getCol(context) == null
-                || CollectionHelper.getInstance().getCol(context).getDecks() == null) {
+        Collection col = getCol(context);
+        if (col == null || col.getDecks() == null) {
             return;
         }
 
@@ -44,6 +44,12 @@ public class BootService extends BroadcastReceiver {
         scheduleNotification(context);
         sWasRun = true;
     }
+
+
+    private Collection getCol(Context context) {
+        return CollectionHelper.getInstance().getCol(context);
+    }
+
 
     private void scheduleDeckReminder(Context context) {
         AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);

--- a/AnkiDroid/src/main/java/com/ichi2/anki/services/BootService.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/services/BootService.java
@@ -17,6 +17,8 @@ import com.ichi2.utils.JSONObject;
 
 import java.util.Calendar;
 
+import timber.log.Timber;
+
 public class BootService extends BroadcastReceiver {
 
     /**
@@ -35,7 +37,7 @@ public class BootService extends BroadcastReceiver {
             return;
         }
         // There are cases where the app is installed, and we have access, but nothing exist yet
-        Collection col = getCol(context);
+        Collection col = getColSafe(context);
         if (col == null || col.getDecks() == null) {
             return;
         }
@@ -46,8 +48,15 @@ public class BootService extends BroadcastReceiver {
     }
 
 
-    private Collection getCol(Context context) {
-        return CollectionHelper.getInstance().getCol(context);
+    private Collection getColSafe(Context context) {
+        //#6239 - previously would crash if ejecting, we don't want a report if this happens so don't use
+        //getInstance().getColSafe
+        try {
+            return CollectionHelper.getInstance().getCol(context);
+        } catch (Exception e) {
+            Timber.e(e, "Failed to get collection for boot service - possibly media ejecting");
+            return null;
+        }
     }
 
 

--- a/AnkiDroid/src/main/java/com/ichi2/anki/services/BootService.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/services/BootService.java
@@ -31,17 +31,21 @@ public class BootService extends BroadcastReceiver {
     @Override
     public void onReceive(Context context, Intent intent) {
         if (sWasRun) {
+            Timber.d("BootService - Already run");
             return;
         }
         if (!Permissions.hasStorageAccessPermission(context)) {
+            Timber.w("Boot Service did not execute - no permissions");
             return;
         }
         // There are cases where the app is installed, and we have access, but nothing exist yet
         Collection col = getColSafe(context);
         if (col == null || col.getDecks() == null) {
+            Timber.w("Boot Service did not execute - error loading collection");
             return;
         }
 
+        Timber.i("Executing Boot Service");
         scheduleDeckReminder(context);
         scheduleNotification(context);
         sWasRun = true;

